009-strimzi-ui.md

```diff
@@ -0,0 +1,262 @@
+
+### Proposed architecture:
+
+![UI session component architecture](./images/009-session-architecture.png)
+
+- Admin – Graphql server, supporting HTTP and websocket connections [admin server proposal](https://github.com/strimzi/proposals/pull/9)
+- Express – UI server, handling sessions for the client
+- Client – Browser running the strimzi UI, executing graphql queries/subscriptions
+- Kafka – brokers (connected to via admin clients)
+- Auth – external auth provider, oauth for Kafka (and possible oauth dance for UI)
+
+### Proposed technology:
+
+The UI is being hosted by an Express server. Express has session middleware - https://www.npmjs.com/package/express-session - that can create and persist server-side sessions, using a cookie as a key to hydrate a session into the incoming express request. Proposed session store is a `redis` container as it's a lightweight key/value (and the session will just be storing a token value).
+
+Authentication in node/Express can be achieved through by http://www.passportjs.org/ which provides a large collection of “strategies” for authenticating a user. For purposes of the proposal, the below diagrams show an oidc flows for a UI oauth dance, although other authentication mechanisms could be used.
+
+### Sequence diagrams:
+
+Note – the log in flow is assuming an oauth dance to retrieve a user token. Depending on the passport strategy and kafka auth mechanism, the UI may need to render its own basic login page.
+
+#### http flow with valid session
+
+![UI http flow with valid session](./images/009-http-valid-session.svg)
+
+#### http flow with session expiry
+
+![UI http flow with session expiry](./images/009-http-session-expiry.svg)
+
+#### websocket flow with valid session
+
+![UI websocket flow with valid session](./images/009-ws-valid-session.svg)
+
+#### websocket handshake flow with session expiry
+
+![UI websocket handshake flow with session expiry](./images/009-ws-handshake-session-expiry.svg)
+
+#### websocket message flow with session expiry
+
+![UI websocket message flow with session expiry](./images/009-ws-message-session-expiry.svg)
+
+## Affected projects
+
+I would expect that the main development effort for a UI will be in a new repository in the Strimzi Github organisation. However, given the discussion in https://github.com/strimzi/strimzi-kafka-operator/issues/2540 , one option to source backend Kafka data would be via the [Bridge](https://github.com/strimzi/strimzi-kafka-bridge). As required, this could be extended to support the various Kafka backend calls needed to surface any required information in a UI. In addition, depending on how and where the UI is provided in Strimzi, changes may be needed in the [strimzi-kafka-operator](https://github.com/strimzi/strimzi-kafka-operator) to either define a UI CRD, or deploy a UI deployment if configured to do so via the operator. I would very much welcome discussion around this, and what would make the most sense for the Strimzi project.
```

Comment:
it sounds to me we are going into the direction of having a dedicated admin API server to avoid adding too many logic to the bridge itself.

Reply:
Agreed - will update this section to reference Strimzi-admin